src/lib/certmap/sss_cert_digest_crypto.c

```diff
@@ -0,0 +1,169 @@
+#include <openssl/evp.h>
+
+#include "lib/certmap/sss_certmap_int.h"
+
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+#define discard_const(ptr) ((void *)((uintptr_t)(ptr)))
+#define EVP_MD_free(ctx)
+#define EVP_MD_fetch(ctx, algorithm, properties) discard_const(EVP_get_digestbyname(algorithm))
+#endif
+
+#define DATA_STEP_SIZE 30
+
+struct get_digest_data {
+    const char **list;
+    size_t size;
+    size_t idx;
+    int error;
+};
+
+static void get_digest_helper(const OBJ_NAME *name, void *arg)
+{
+    struct get_digest_data *data = (struct get_digest_data *) arg;
+    EVP_MD *md = NULL;
+
+    if (data->error != 0) {
+        return;
+    }
+
+    if (name == NULL || name->name == NULL || islower(*name->name) == 0
+        || strstr(name->name, "RSA") != NULL
+        || strstr(name->name, "rsa") != NULL) {
+        return;
+    }
+
+    md = EVP_MD_fetch(NULL, name->name, NULL);
+    if (md != NULL) {
+        data->list[data->idx] = talloc_strdup(data->list, name->name);
+        if (data->list[data->idx] == NULL) {
+            data->error = ENOMEM;
+            return;
+        } else {
+            data->idx++;
+            if (data->idx == data->size - 1) {
+                data->size += DATA_STEP_SIZE;
+                data->list = talloc_realloc(data->list, data->list,
+                                            const char *, data->size);
+                if (data->list == NULL) {
+                    data->error = ENOMEM;
+                    return;
+                }
+            }
+            data->list[data->idx] = NULL;
+        }
+    }
+    EVP_MD_free(md);
+}
+
+
+int get_digest_list(TALLOC_CTX *mem_ctx, const char ***digest_list)
+{
+    struct get_digest_data data = { 0 };
+
+    data.size = DATA_STEP_SIZE;
+    data.list = talloc_array(mem_ctx, const char *, data.size);
+    if (data.list == NULL) {
+        return ENOMEM;
+    }
+
+    OpenSSL_add_all_algorithms();
+
+    OBJ_NAME_do_all(OBJ_NAME_TYPE_MD_METH, get_digest_helper, &data);
+    if (data.error != 0) {
+        talloc_free(data.list);
+        return data.error;
+    }
+
+    *digest_list = data.list;
+
+    return 0;
+}
+
+int get_hash(TALLOC_CTX *mem_ctx, const uint8_t *blob, size_t blob_size,
+             const char *digest, bool upper, bool colon, bool reverse,
+             char **out)
+{
+    int ret;
+    EVP_MD *md = NULL;
+    EVP_MD_CTX *md_ctx = NULL;
+    unsigned char md_value[EVP_MAX_MD_SIZE];
+    unsigned int md_len;
+    char *tmp_str = NULL;
+
+    md = EVP_MD_fetch(NULL, digest, NULL);
+    if (md == NULL) {
+        return EINVAL;
+    }
+
+    md_ctx = EVP_MD_CTX_create();
+    if (md_ctx == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = EVP_DigestInit(md_ctx, md);
+    if (ret != 1) {
+        ret = EIO;
+        goto done;
+    }
+
+    ret = EVP_DigestUpdate(md_ctx, blob, blob_size);
+    if (ret != 1) {
+        ret = EIO;
+        goto done;
+    }
+
+    ret = EVP_DigestFinal_ex(md_ctx, md_value, &md_len);
```

Comment:
Why not single-shot `EVP_Digest()` instead?

Reply:
Replaced